src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```
@@ -0,0 +1,510 @@
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            var awsCredentialsCreators = new Func<AwsCredentials>[]
+            {
+                () => CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => CreateAwsCredentialsFromEnvironmentVariables(),
+                () => CreateAwsCredentialsFromEcsResponse(),
+                () => CreateAwsCredentialsFromEc2Response()
+            };
+
+            foreach (var awsCredentialsCreator in awsCredentialsCreators)
+            {
+                var awsCredentials = awsCredentialsCreator();
+
+                if (awsCredentials == null)
+                {
+                    continue;
+                }
+
+                ValidateAwsCredentials(awsCredentials);
+
+                if (awsCredentials.Username == null)
```

Why would Username ever be null?

If having a null Username is not valid then the Create method should have returned null or thrown an exception instead of returning an invalid awsCredentials.

---

Done.